File: Sudoku_Java_Shell/src/ConstraintNetwork.java
import java.util.*;
import java.util.Map.*;

/**
 * CSP representation of the problem. Contains the variables, constraints, and
 * many helpful accessors.
 */

public class ConstraintNetwork
{

	// =================================================================
	// Properties
	// =================================================================

	private List<Constraint> constraints;
	private List<Variable> variables;

	// =================================================================
	// Constructors
	// =================================================================

	public ConstraintNetwork ( )
	{
		variables = new ArrayList<Variable>();
		constraints = new ArrayList<Constraint>();
	}

	public ConstraintNetwork ( SudokuBoard sboard )
	{
		int [][] board = sboard.getBoard();

		List<Variable> temp = new ArrayList<Variable>();
		int value = 0;

		for ( int i = 0; i < sboard.getN(); i++ )
		{
			for ( int j = 0; j < sboard.getN(); j++ )
			{
				value = board[i][j];
				List<Integer> domain = new ArrayList<Integer>();

				if (value == 0)
				{
					for (int d = 1; d <= sboard.getN(); d++)
					{
						domain.add(d);
					}
				}
				else
				{
					domain.add(value);
				}

				int block = (i/sboard.getP() * sboard.getP()) + j/sboard.getQ();

				temp.add(new Variable(domain, i, j, block));
			}
		}

		Map<Integer, List<Variable>> rows = new HashMap<Integer, List<Variable>>();
		Map<Integer, List<Variable>> cols = new HashMap<Integer, List<Variable>>();
		Map<Integer, List<Variable>> blocks = new HashMap<Integer, List<Variable>>();

		for ( Variable v : temp )
		{
			int row = v.row();
			int col = v.col();
			int block = v.block();

			if (!rows.containsKey(row))
			{
				rows.put(row, new ArrayList<Variable>());
			}

			if (!cols.containsKey(col))
			{
				cols.put(col, new ArrayList<Variable>());
			}

			if (!blocks.containsKey(block))
			{
				blocks.put(block, new ArrayList<Variable>());
			}
			rows.get(row).add(v);
			cols.get(col).add(v);
			blocks.get(block).add(v);
		}

		variables = new ArrayList<Variable>();
		constraints = new ArrayList<Constraint>();

		for ( Variable v : temp )
		{
			add(v);
		}

		for ( Entry<Integer, List<Variable>> e : rows.entrySet() )
		{
			Constraint c = new Constraint();
			for ( Variable v : e.getValue() )
			{
				c.addVariable(v);
			}
			add(c);
		}

		for ( Entry<Integer, List<Variable>> e : cols.entrySet() )
		{
			Constraint c = new Constraint();
			for ( Variable v : e.getValue() )
			{
				c.addVariable(v);
			}
			add(c);
		}

		for ( Entry<Integer, List<Variable>> e : blocks.entrySet() )
		{
			Constraint c = new Constraint();
			for ( Variable v : e.getValue() )
			{
				c.addVariable(v);
			}
			add(c);
		}
	}

	// =================================================================
	// Modifiers
	// =================================================================

	public void add ( Constraint c )
	{
		if ( ! constraints.contains( c ) )
		{
			constraints.add( c );
		}
	}

	public void add ( Variable v )
	{
		if ( ! variables.contains( v ) )
		{
			variables.add( v );
		}
	}

	// =================================================================
	// Accessors
	// =================================================================

	public List<Constraint> getConstraints ( )
	{
		return constraints;
	}

	public List<Variable> getVariables ( )
	{
		return variables;
	}

	// Returns all variables that share a constraint with v
	public List<Variable> getNeighborsOfVariable ( Variable v )
	{
		Set<Variable> neighbors = new HashSet<Variable>();

		for ( Constraint c : constraints )
		{
			if ( c.contains( v ) )
			{
				neighbors.addAll( c.vars );
			}
		}

		neighbors.remove( v );
		return new ArrayList<Variable>( neighbors );
	}

	// Returns true is every constraint is consistent
	public boolean isConsistent ( )
	{
		for ( Constraint c : constraints )
			if ( ! c.isConsistent() )
				return false;

		return true;
	}

	// Returns a list of constraints that contains v
	public List<Constraint> getConstraintsContainingVariable ( Variable v )
	{
		List<Constraint> outList = new ArrayList<Constraint>();

		for ( Constraint c : constraints )
			if ( c.contains( v ) )
				outList.add( c );

		return outList;
	}

	/**
	 * Returns the constraints that contain variables whose domains were
	 * modified since the last call to this method.
	 *
	 * After getting the constraints, it will reset each variable to
	 * unmodified.
	 *
	 * Note* The first call to this method returns the constraints containing
	 * the initialized variables.
	 */
	public List<Constraint> getModifiedConstraints ( )
	{
		List<Constraint> mConstraints = new ArrayList<Constraint>();

		for ( Constraint c : constraints )
			if ( c.isModified() )
				mConstraints.add( c );

		for ( Variable v : getVariables() )
			v.setModified( false );

		return mConstraints;
	}

	// =================================================================
	// String Representation
	// =================================================================

	public String toString ( )
	{
		StringBuilder sb = new StringBuilder();
		sb.append( variables.size() + " Variables: {" );
		String delim = "";

		for ( Variable v : variables )
		{
			sb.append( delim ).append( v.getName() );
			delim = ",";
		}
		sb.append( "}" );

		sb.append( "\n" + constraints.size() + " Constraints:" );
		delim = "\n";
		for ( Constraint c : constraints )
		{
			sb.append( delim ).append( c.toString() );
		}
		return sb.toString();
	}

	// =================================================================
	// Sudoku Board Representation
	// =================================================================

	public SudokuBoard toSudokuBoard ( int p, int q )
	{
		int n = p*q;
		int [][] board = new int[n][n];
		int row = 0, col = 0;
		for ( Variable v : getVariables() )
		{
			board[row][col++] = v.getAssignment();
			if ( col == n )
			{
				col = 0;
				row++;
			}
		}
		return new SudokuBoard( p, q, board );
	}
}

File: Sudoku_Cpp_Shell/src/ConstraintNetwork.cpp
#include "ConstraintNetwork.hpp"

/**
 * CSP representation of the problem. Contains the variables, constraints, and
 * many helpful accessors.
 */

// =====================================================================
// Constructors
// =====================================================================

ConstraintNetwork::ConstraintNetwork ( void )
{

}

ConstraintNetwork::ConstraintNetwork ( SudokuBoard sboard )
{
	SudokuBoard::Board board = sboard.get_board();

	VariableSet variables;
	int value = 0;
	int N = sboard.get_p() * sboard.get_q();

	for ( int i = 0; i < N; ++i )
	{
		for ( int j = 0; j < N; ++j )
		{
			value = board[i][j];
			std::vector<int> domain;
			if ( value == 0 )
			{
				for ( int d = 1; d <= N; ++d )
				{
					domain.push_back( d );
				}
			}
			else
			{
				domain.push_back( value );
			}

			int block = ( i / sboard.get_p() * sboard.get_p() + j / sboard.get_q() );

			Variable* toPushBack = new Variable(domain,i,j,block);
			variables.push_back(toPushBack);
		}
	}

	std::map<int, VariableSet > rows;
	std::map<int, VariableSet > cols;
	std::map<int, VariableSet > blocks;

	for ( Variable* v:variables )
	{
		int row = v->row();
		int col = v->col();
		int block = v->block();

		if ( !rows.count( row ) )
		{
			rows[row] = VariableSet();
		}
		if ( !cols.count( col ) )
		{
			cols[col] = VariableSet();
		}
		if ( !blocks.count( block ) )
		{
			blocks[block] = VariableSet();
		}
		rows[row].push_back(v);
		cols[col].push_back(v);
		blocks[block].push_back(v);
	}

	for ( Variable* v : variables )
	{
		add( v );
	}

	for ( auto iterator = rows.begin(); iterator != rows.end(); ++iterator )
	{
		Constraint c = Constraint();
		for ( Variable* vv:iterator->second )
		{
			c.addVariable(vv);
		}

		add( c );
	}

	for ( auto iterator = cols.begin(); iterator != cols.end(); ++iterator)
	{
		 Constraint c = Constraint();
		 for ( Variable* vv:iterator->second )
		 {
			c.addVariable(vv);
		 }
		 add( c );
	 }

	 for ( auto iterator = blocks.begin(); iterator != blocks.end(); ++iterator)
	 {
		 Constraint c = Constraint();
		 for ( Variable* vv:iterator->second )
		 {
			 c.addVariable(vv);
		 }
		 add( c );
	  }
}

// =====================================================================
// Modifiers
// =====================================================================

void ConstraintNetwork::add ( Constraint c )
{
	if ( std::find(constraints.begin(),constraints.end(),c) == constraints.end()){
		constraints.push_back( c );
	}
}

void ConstraintNetwork::add ( Variable* v )
{
	if ( std::find( variables.begin(), variables.end(), v ) == variables.end() )
	{
		variables.push_back( v );
	}
}

// =====================================================================
// Accessors
// =====================================================================

ConstraintNetwork::ConstraintSet ConstraintNetwork::getConstraints ( void ) const
{
	return constraints;
}

ConstraintNetwork::VariableSet ConstraintNetwork::getVariables ( void ) const
{
	return variables;
}

// Returns all variables that share a constraint with v
ConstraintNetwork::VariableSet ConstraintNetwork::getNeighborsOfVariable ( Variable* v )
{
	ConstraintNetwork::VariableSet neighbors;

	for ( Constraint& c : constraints )
	{
		if ( c.contains( v ) )
		{
			for ( Variable* va : c.vars )
				if ( va != v && std::find(neighbors.begin(),neighbors.end(),va) == neighbors.end() )
					neighbors.push_back( va );
		}
	}

	return neighbors;
}

// Returns true is every constraint is consistent
bool ConstraintNetwork::isConsistent ( void ) const
{
	for ( Constraint c : constraints )
		if ( ! c.isConsistent() )
			return false;

	return true;
}

// Returns a list of constraints that contains v
ConstraintNetwork::ConstraintRefSet ConstraintNetwork::getConstraintsContainingVariable ( Variable* v )
{
	ConstraintNetwork::ConstraintRefSet outList;

	for ( Constraint& c : constraints )
		if ( c.contains( v ) )
			outList.push_back( &c );

	return outList;
}

/**
 * Returns the constraints that contain variables whose domains were
 * modified since the last call to this method.
 *
 * After getting the constraints, it will reset each variable to
 * unmodified.
 *
 * Note* The first call to this method returns the constraints containing
 * the initialized variables.
 */
ConstraintNetwork::ConstraintRefSet ConstraintNetwork::getModifiedConstraints ( void )
{
	ConstraintNetwork::ConstraintRefSet mConstraints;

	for ( Constraint& c : constraints )
		if ( c.isModified() )
			mConstraints.push_back( &c );

	for ( Variable* v : variables )
		v->setModified( false );

	return mConstraints;
}

// =====================================================================
// String Representation
// =====================================================================

std::string ConstraintNetwork::toString ( void ) const
{
	std::stringstream ss;
	ss << variables.size() << " Variables: {";
	std::string delim = "";

	for ( Variable* v : variables )
	{
		ss << delim << v->getName();
		delim = ",";
	}
	ss<<"}";

	ss << "\n" << constraints.size() << " Constraints:";
	delim = "\n";
	for ( Constraint c: constraints )
	{
		ss << delim << c.toString();
	}
	return ss.str();
}

// =====================================================================
// Sudoku Board Representation
// =====================================================================

SudokuBoard ConstraintNetwork::toSudokuBoard ( int _p, int _q ) const
{
	SudokuBoard::Board board;
	int n = _p*_q;

	for ( int i = 0; i < n; ++i )
	{
		SudokuBoard::Row row;

		for ( int j = 0; j < n; ++j )
			row.push_back( 0 );

		board.push_back( row );
	}

	int row = 0, col = 0;
	for ( Variable* v: getVariables() )
	{
		board[row][col++] = v->getAssignment();

		if ( col >= n )
		{
			col = 0;
			row++;
		}
	}

	return SudokuBoard( _p, _q, board );
}
